reactor-core/src/main/java/reactor/util/Metrics.java

```diff
@@ -49,4 +63,72 @@ public static final boolean isInstrumentationAvailable() {
 		return isMicrometerAvailable;
 	}
 
+
+	/**
+	 * If the Micrometer instrumentation facade is available, return a {@link BiFunction}
+	 * that takes a {@link Scheduler} and a {@link ScheduledExecutorService} to instrument.
+	 * It can be applied to instruments {@link ExecutorService}-based schedulers (as
+	 * supported by Micrometer, ie. it instruments state of queues but not timing of tasks).
+	 * <p>
+	 * This factory sends instrumentation data to the Micrometer Global Registry.
+	 *
+	 * @implNote Note that this is added as a decorator via Schedulers when enabling metrics for schedulers, which doesn't change the Factory.
+	 *
+	 * @return a {@link BiFunction} that decorates {@link ScheduledExecutorService} to instrument their internal state,
+	 * or is the identity function if Micrometer isn't available.
+	 */
+	public static BiFunction<Scheduler, ScheduledExecutorService, ScheduledExecutorService> schedulersDecorator() {
+		if (isMicrometerAvailable) {
+			return new MicrometerSchedulersDecorator();
+		}
+		return (scheduler, service) -> service;
+	}
+
+	static final class MicrometerSchedulersDecorator
+			implements BiFunction<Scheduler, ScheduledExecutorService, ScheduledExecutorService> {
+
+		final Map<String, AtomicInteger>                                             SCHEDULERS_NAME_COUNTER = new HashMap<>();
+		final Map<String, WeakHashMap<Scheduler, String>>                            UNIQUE_SCHEDULER_NAMES = new HashMap<>();
+
+		final WeakHashMap<Scheduler, AtomicInteger>                                  EXECUTORS_NAME_COUNTER = new WeakHashMap<>();
+		final WeakHashMap<Scheduler, WeakHashMap<ScheduledExecutorService, String>>  EXECUTOR_LOCAL_IDS     = new WeakHashMap<>();
+
+		@Override
+		public synchronized ScheduledExecutorService apply(Scheduler scheduler, ScheduledExecutorService service) {
+			String schedulerName = Scannable
+					.from(scheduler)
+					.scanOrDefault(Attr.NAME, scheduler.getClass().getName());
+
+			String schedulerId = UNIQUE_SCHEDULER_NAMES
+					.computeIfAbsent(schedulerName, key -> new WeakHashMap<>())
+					.computeIfAbsent(scheduler, key -> {
+						int seqNum = SCHEDULERS_NAME_COUNTER
+								.computeIfAbsent(schedulerName, nameKey -> new AtomicInteger())
+		                        .getAndIncrement();
+
+						if (seqNum > 0) {
+							// If we got a name clash, append a sequential number to the name
+							return schedulerName + "#" + seqNum;
+						}
+						else {
+							return schedulerName;
+						}
+					});
+
+			String executorName = EXECUTOR_LOCAL_IDS
```

Comment:
shouldn't the executorName be keyed on the previously computed unique `schedulerId`? Plus it wouldn't need a second `WeakHashMap<Scheduler,...>` this way?

Reply:
Sure, we can key it by `schedulerId`

Regarding "plus" - we still need `WeakHashMap<Scheduler,...>` to keep it same between invocations